File: Problems/Algorithms/581. Shortest Unsorted Continuous Subarray/ShortestSubarray.java
class Solution {
    public int findUnsortedSubarray(int[] nums) {
        int n = nums.length;
        Deque<Integer> stack = new ArrayDeque<>();
        
        int left = n-1;
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && nums[i] < nums[stack.peek()]) {
                left = Math.min(left, stack.pop());
            }
            stack.push(i);
        }
        
        stack.clear();
        
        int right = 0;
        for (int i = n-1; i >= 0; i--) {
            while (!stack.isEmpty() && nums[i] > nums[stack.peek()]) {
                right = Math.max(right, stack.pop());
            }
            stack.push(i);
        }
        
        return right-left > 0 ? right-left+1 : 0;
    }
}

File: Problems/Algorithms/581. Shortest Unsorted Continuous Subarray/ShortestSubarray.kt
class Solution {
    fun findUnsortedSubarray(nums: IntArray): Int {
        val newNums = nums.copyOf()
        newNums.sort()
        
        var i = 0
        var j = nums.size - 1
        while (i < j) {
            if (nums[i] != newNums[i] && nums[j] != newNums[j]) {
                return j-i+1
            }
            
            if (nums[i] == newNums[i]) {
                i++
            }
            
            if (nums[j] == newNums[j]) {
                j--
            }
        }
        
        return 0
    }
}

